src/Isbn.php

```diff
@@ -61,22 +67,34 @@ private static function extractIsbnAs($str)
 
     private static function extractIsbn13s($str)
     {
-        preg_match_all(self::ISBN_13_REGEXP, $str, $matches);
+        preg_match_all(self::ISBN_13_REGEXP, $str, $matches, \PREG_SET_ORDER);
 
         return array_filter(
-            array_map([__CLASS__, 'stripHyphenation'], $matches[0]),
+            array_map(
+                [__CLASS__, 'stripHyphenation'],
+                array_filter(
+                    $matches,
+                    [__CLASS__, 'isValidIsbn13Grouping']
```

Comment:
Instead of using private static methods, we _could_ experiment with using [classes that respond to `__invoke` as they are also considered `callable`](http://php.net/manual/en/language.types.callable.php), e.g.

```php
array_filter($matches, new IsbnGroupFilter(4));

// ...

class IsbnGroupFilter
{
    private $number;

    public function __construct($number)
    {
        $this->number = $number;
    }

    public function __invoke($match)
    {
        if (empty($match[2])) {
            return true;
        }

        return mb_substr_count($match[1], $match[2]) === $this->number;
    }
}
```

I'm not sure this is any more maintainable though.

Reply:
For comparison, here's the same code expanded to use anonymous functions:

```php
return array_filter(
    array_map(
        function ($match) {
            if (empty($match[2])) {
                return $match[1];
            }

            return str_replace($match[2], '', $match[1]);
        },
        array_filter(
            $matches,
            function ($match) {
                if (empty($match[2])) {
                    return true;
                }

                return mb_substr_count($match[1], $match[2]) === 4;
            }
        )
    ),
    [__CLASS__, 'isValidIsbn13']
);
```

Reply:
And here's another option, ditching the mix of functional and object-oriented style:

```php
    private static function extractIsbn13s($str)
    {
        preg_match_all(self::ISBN_13_REGEXP, $str, $matches, \PREG_SET_ORDER);

        $isbns = [];

        foreach ($matches as $match) {
            list($_, $isbn, $hyphen) = array_pad($match, 3, null);

            if ($hyphen) {
                if (mb_substr_count($isbn, $hyphen) !== 4) {
                    continue;
                }

                $isbn = str_replace($hyphen, '', $isbn);
            }

            if (self::isValidIsbn13($isbn)) {
                $isbns[] = $isbn;
            }
        }

        return $isbns;
    }
```